File: bms/go/triangolatte/Polygon/Polygon.go
package test

import (
	"errors"
	"math"
)

var ()

const ()

func Polygon(points []Point) ([]float64, error) {
	n := len(points)

	if n < 3 {
		return nil, errors.New("cannot triangulate less than three points")
	}

	// Allocate memory for all needed elements and initialize them by hand.
	elements := make([]Element, n)
	elements[0].Prev, elements[0].Next = &elements[n-1], &elements[1]
	elements[0].Point = points[0]
	for i := 1; i < n-1; i++ {
		elements[i].Prev, elements[i].Next = &elements[i-1], &elements[i+1]
		elements[i].Point = points[i]
	}
	elements[n-1].Prev, elements[n-1].Next = &elements[n-2], &elements[0]
	elements[n-1].Point = points[n-1]

	ear := &elements[0]

	// Any triangulation of simple polygon has n-2 triangles. Triangle has 3
	// two-dimensional coordinates.
	i, t := 0, make([]float64, (n-2)*6)

	stop := ear
	var prev, next *Element

	for ear.Prev != ear.Next {
		prev = ear.Prev
		next = ear.Next

		if isEar(ear) {
			if polygonArea([]Point{prev.Point, ear.Point, next.Point}) > 0 {
				t[i+0], t[i+1] = prev.Point.X, prev.Point.Y
				t[i+2], t[i+3] = ear.Point.X, ear.Point.Y
				t[i+4], t[i+5] = next.Point.X, next.Point.Y
				i += 6
			}

			ear.Remove()
			ear = ear.Next
			stop = stop.Next
			continue
		}

		ear = next

		if ear == stop {
			return []float64{}, errors.New("oops")
		}
	}

	// Return array slice of size consisting only of the elements actually took by
	// the triangulation (sometimes the number of triangles is lower than n-2 and
	// zeroes are filling the rest of the array).
	return t[0:i], nil
}

func isEar(p *Element) bool {
	a, b, c := p.Prev.Point, p.Point, p.Next.Point
	if isReflex(a, b, c) {
		return false
	}

	r := p.Next.Next
	for r != p.Prev {
		inside := isInsideTriangle(a, b, c, r.Point)
		reflex := isReflex(r.Prev.Point, r.Point, r.Next.Point)
		if inside && reflex {
			return false
		}
		r = r.Next
	}
	return true
}

func isReflex(a, b, c Point) bool {
	return (b.X-a.X)*(c.Y-b.Y)-(c.X-b.X)*(b.Y-a.Y) < 0
}

func isInsideTriangle(a, b, c, p Point) bool {
	return (c.X-p.X)*(a.Y-p.Y)-(a.X-p.X)*(c.Y-p.Y) >= 0 &&
		(a.X-p.X)*(b.Y-p.Y)-(b.X-p.X)*(a.Y-p.Y) >= 0 &&
		(b.X-p.X)*(c.Y-p.Y)-(c.X-p.X)*(b.Y-p.Y) >= 0
}

func polygonArea(data []Point) float64 {
	area := 0.0
	for i, j := 0, len(data)-1; i < len(data); i++ {
		area += data[i].X*data[j].Y - data[i].Y*data[j].X
		j = i
	}
	return math.Abs(area / 2)
}

func (e *Element) Remove() {
	e.Next.Prev = e.Prev
	e.Prev.Next = e.Next
}

type Point struct {
	X, Y float64
}

type Element struct {
	Prev, Next *Element
	Point      Point
}

File: translations_only/success/go/triangolatte/Polygon/gpt4/Hinted/Polygon.rs

use std::error::Error;
use std::f64;

#[derive(Clone, Copy)]
struct Point {
    x: f64,
    y: f64,
}

struct Element {
    prev: usize,
    next: usize,
    point: Point,
}

fn polygon(points: Vec<Point>) -> Result<Vec<f64>, Box<dyn Error>> {
    let n = points.len();

    if n < 3 {
        return Err("cannot triangulate less than three points".into());
    }

    let mut elements: Vec<Element> = Vec::with_capacity(n);
    elements.push(Element {
        prev: n - 1,
        next: 1,
        point: points[0],
    });
    for i in 1..n - 1 {
        elements.push(Element {
            prev: i - 1,
            next: i + 1,
            point: points[i],
        });
    }
    elements.push(Element {
        prev: n - 2,
        next: 0,
        point: points[n - 1],
    });

    let mut ear = 0;
    let mut t: Vec<f64> = vec![0.0; (n - 2) * 6];
    let mut i = 0;

    let mut stop = ear;
    let mut prev;
    let mut next;

    while elements[ear].prev != elements[ear].next {
        prev = elements[ear].prev;
        next = elements[ear].next;

        if is_ear(&elements, ear) {
            if polygon_area(vec![elements[prev].point, elements[ear].point, elements[next].point]) > 0.0 {
                t[i] = elements[prev].point.x;
                t[i + 1] = elements[prev].point.y;
                t[i + 2] = elements[ear].point.x;
                t[i + 3] = elements[ear].point.y;
                t[i + 4] = elements[next].point.x;
                t[i + 5] = elements[next].point.y;
                i += 6;
            }

            remove(&mut elements, ear);
            ear = next;
            stop = elements[stop].next;
            continue;
        }

        ear = next;

        if ear == stop {
            return Err("oops".into());
        }
    }

    t.truncate(i);
    Ok(t)
}

fn is_ear(elements: &Vec<Element>, p: usize) -> bool {
    let a = elements[elements[p].prev].point;
    let b = elements[p].point;
    let c = elements[elements[p].next].point;
    if is_reflex(a, b, c) {
        return false;
    }

    let mut r = elements[elements[p].next].next;
    while r != elements[p].prev {
        let inside = is_inside_triangle(a, b, c, elements[r].point);
        let reflex = is_reflex(elements[elements[r].prev].point, elements[r].point, elements[elements[r].next].point);
        if inside && reflex {
            return false;
        }
        r = elements[r].next;
    }
    true
}

fn is_reflex(a: Point, b: Point, c: Point) -> bool {
    (b.x - a.x) * (c.y - b.y) - (c.x - b.x) * (b.y - a.y) < 0.0
}

fn is_inside_triangle(a: Point, b: Point, c: Point, p: Point) -> bool {
    (c.x - p.x) * (a.y - p.y) - (a.x - p.x) * (c.y - p.y) >= 0.0 &&
    (a.x - p.x) * (b.y - p.y) - (b.x - p.x) * (a.y - p.y) >= 0.0 &&
    (b.x - p.x) * (c.y - p.y) - (c.x - p.x) * (b.y - p.y) >= 0.0
}

fn polygon_area(data: Vec<Point>) -> f64 {
    let mut area = 0.0;
    let mut j = data.len() - 1;
    for i in 0..data.len() {
        area += data[i].x * data[j].y - data[i].y * data[j].x;
        j = i;
    }
    area.abs() / 2.0
}

fn remove(elements: &mut Vec<Element>, index: usize) {
    let prev = elements[index].prev;
    let next = elements[index].next;
    elements[prev].next = next;
    elements[next].prev = prev;
}
